service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/serializers/ParserUtility.java

```diff
@@ -0,0 +1,499 @@
+    {
+        /* Codes_SRS_PARSER_UTILITY_21_048: [The validateMap shall do nothing if the map is null.] */
+        if(map != null)
+        {
+            /* Codes_SRS_PARSER_UTILITY_21_047: [The validateMap shall do nothing if the map is a valid Map.] */
+            validateMapInternal(map);
+        }
+    }
+
+    private static void validateMapInternal(Map<String, Object> map) throws IllegalArgumentException
+    {
+        for(Map.Entry<String, Object> entry : map.entrySet())
+        {
+            Object value = entry.getValue();
+
+            if ((value != null) && ((value.getClass().isArray()) || (value.getClass().isLocalClass())))
+            {
+                throw new IllegalArgumentException("Map contains illegal value type " + value.getClass().getName());
+            }
+
+
+            if (value instanceof Map)
+            {
+                validateMapInternal((Map<String, Object>) value);
+            }
+        }
+    }
+
+    /**
+     * Validate if a provided ID is valid using the follow criteria.
+     * A case-sensitive string (up to 128 char long)
+     * of ASCII 7-bit alphanumeric chars
+     * + {'-', ':', '.', '+', '%', '_', '#', '*', '?', '!', '(', ')', ',', '=', '@', ';', '$', '''}.
+     *
+     * @param id is the ID to test
+     * @throws IllegalArgumentException if the ID do not fits the criteria
+     */
+    public static void validateId(String id) throws IllegalArgumentException
+    {
+        /* Codes_SRS_PARSER_UTILITY_21_026: [The validateId shall throw IllegalArgumentException if the provided string is null or empty.] */
+        /* Codes_SRS_PARSER_UTILITY_21_027: [The validateId shall throw IllegalArgumentException if the provided string contains at least one not UTF-8 character.] */
+        try
+        {
+            validateStringUTF8(id);
+        }
+        catch (IllegalArgumentException e)
+        {
+            throw new IllegalArgumentException("The provided ID is not valid");
+        }
+
+        /* Codes_SRS_PARSER_UTILITY_21_028: [The validateId shall throw IllegalArgumentException if the provided string contains more than 128 characters.] */
+        if(id.length() > 128)
+        {
+            throw new IllegalArgumentException("The provided ID is bigger than 128 characters");
+        }
+
+        /* Codes_SRS_PARSER_UTILITY_21_029: [The validateId shall throw IllegalArgumentException if the provided string contains an illegal character.] */
+        byte[] chars = id.getBytes(StandardCharsets.UTF_8);
+        for (byte c:chars)
+        {
+            if(!(((c>='A') && (c<='Z')) || ((c>='a') && (c<='z')) || ((c>='0') && (c<='9')) ||
+                    (c=='-') || (c==':') || (c=='.') || (c=='+') || (c=='%') || (c=='_') || (c=='#') || (c=='*') || (c=='?') ||
+                    (c=='!') || (c=='(') || (c==')') || (c==',') || (c=='=') || (c=='@') || (c==';') || (c=='$') || (c=='\'')))
```

Comment:
Given it is a 128-character ID byte array, what about using string matches?

String s = Character.toString(c);
if (s.matches("[0-9a-zA-Z!@#$%?\\\\+\\-\\*\\.:;<()|_]+"))

Reply:
Actually, we don't even need this function anymore. I'll delete it